sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD.scala

```diff
@@ -39,7 +39,19 @@ private[sql] object JsonRDD extends Logging {
       json: RDD[String],
       schema: StructType,
       columnNameOfCorruptRecords: String): RDD[Row] = {
-    parseJson(json, columnNameOfCorruptRecords).map(parsed => asRow(parsed, schema))
+    // Reuse the mutable row for each record, however we still need to 
+    // create a new row for every nested struct type in each record
+    val mutableRow = new SpecificMutableRow(schema.fields.map(_.dataType))
+    parseJson(json, columnNameOfCorruptRecords).mapPartitions( iter => {
+      iter.map { parsed =>
+        schema.fields.zipWithIndex.foreach {
```

Comment:
BTW, I believe that `schema.fields.zipWithIndex` will create temporal objects, use for loop instead.
```
for (i <- 0 until schema.fields.length) {
  val fname = schema.fields(i).name
  val ftype = schema.fields(i).dataType

  mutableRow(i) = ...
}
```